File: 2014-07-18-FP/RemoveDuplicates/rd.accepted.clj
(require 'clojure.string)

(defn f [acc c]
    (if (contains? (first acc) c)
        acc
        (cons (conj (first acc) c) (cons c (rest acc)))))

(defn nb [str]
    (clojure.string/join (reverse (rest (reduce f (cons #{} "") str)))))

(println (nb (read-line)))

File: 2014-07-18-FP/RemoveDuplicates/rd.accepted.scala
object Solution {
    def nb(str: String): String = {
        def f(acc: (Set[Char], List[Char]), c: Char): (Set[Char], List[Char]) = {
            if (acc._1 contains c)
                acc
            else
                (acc._1 + c, c :: acc._2)
        }

        (str foldLeft ((Set[Char](), List[Char]())))(f)._2.reverse mkString ""
    }

    def main(args: Array[String]) = {
        val str = readLine
        println(nb(str))
    }
}
